src/main/java/frc/robot/subsystems/Elevator.java

```diff
@@ -0,0 +1,156 @@
+/*----------------------------------------------------------------------------*/
+/* Copyright (c) 2018-2019 FIRST. All Rights Reserved.                        */
+/* Open Source Software - may be modified and shared by FRC teams. The code   */
+/* must be accompanied by the FIRST BSD license file in the root directory of */
+/* the project.                                                               */
+/*----------------------------------------------------------------------------*/
+
+package frc.robot.subsystems;
+import com.ctre.phoenix.motorcontrol.ControlMode;
+import com.ctre.phoenix.motorcontrol.can.WPI_TalonFX;
+
+import edu.wpi.first.math.MathUtil;
+import edu.wpi.first.math.controller.PIDController;
+import edu.wpi.first.wpilibj.DigitalInput;
+import edu.wpi.first.wpilibj.DutyCycleEncoder;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.constants.Constants;
+import frc.robot.constants.ElevatorConstants;
+import frc.robot.util.MotorFactory;
+import lib.ctre_shims.TalonEncoder;
+
+public class Elevator extends SubsystemBase {
+
+  private final WPI_TalonFX m_motor;
+  private final DigitalInput m_bottomLimitSwitch;
+  private final DigitalInput m_topLimitSwitch;
+  private final PIDController m_elevatorPID;
+  private final DutyCycleEncoder m_absoluteSpoolEncoder;     
+  private final TalonEncoder m_elevatorMotorEncoder; 
+  private double m_absEncoderZeroPositionTicks;
+  private double clampLow = -ElevatorConstants.kMotorLimit; 
+  private double clampHigh = ElevatorConstants.kMotorLimit;
+  private boolean m_enabled; 
+
+  public Elevator() {
+    m_motor = MotorFactory.createTalonFX(ElevatorConstants.kMotorPort, Constants.kRioCAN);
+    //m_motor.configForwardLimitSwitchSource(LimitSwitchSource.FeedbackConnector, LimitSwitchNormal.NormallyOpen, 0);
+    //m_motor.configReverseLimitSwitchSource(LimitSwitchSource.FeedbackConnector, LimitSwitchNormal.NormallyOpen, 0);
+    m_bottomLimitSwitch = new DigitalInput(ElevatorConstants.kBottomLimitSwitchPort); 
+    m_topLimitSwitch = new DigitalInput(ElevatorConstants.kTopLimitSwitchPort);
+    
+    m_absoluteSpoolEncoder = new DutyCycleEncoder(ElevatorConstants.kAbsEncoderPort); 
+    
+    m_elevatorMotorEncoder = new TalonEncoder(m_motor); 
+    m_elevatorMotorEncoder.setDistancePerPulse(ElevatorConstants.kDistPerMotorEncoderTick);
+
+    m_elevatorPID = new PIDController(ElevatorConstants.kP, ElevatorConstants.kI, ElevatorConstants.kD);  
+  }
+
+  public void setMotorLimit(double powerLevel){
+    clampLow = -powerLevel; 
+    clampHigh = powerLevel; 
+  }
+
+  public void setMotorLimit(){
+    setMotorLimit(ElevatorConstants.kMotorLimit);
+  }
+
+  @Override
+  public void periodic() {
+    /**
+     * If we hit the bottom limit switch, then we must be at the zero position.
+     * Thus set the setpoint to 0, then atSetpoint() will be triggered, causing the 
+     * ResetEncoderAtBottom() command to end
+     */
+    if(getBottomLimitSwitch()){
```

Comment:
This would prevent the motor from ever powering upward once it hits the bottom limit switch. I think we talked about this in class. You can get the motor power with motor.get() and if it is positive, then it does not have to be set to zero.
But, I would prefer you use the built in TalonFX limit switch trigger to stop the motor.

Reply:
I have used the built in TalonFX limit switch trigger to stop the motor. I may or may not have done it right.